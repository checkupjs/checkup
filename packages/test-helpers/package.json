{
  "name": "@checkup/test-helpers",
  "description": "Checkup's test helper library",
  "version": "0.0.10",
  "author": "Steve Calvert <steve.calvert@gmail.com>",
  "bugs": "https://github.com/checkupjs/checkup/issues",
  "dependencies": {
    "@checkup/core": "0.0.10",
    "fixturify-project": "^2.1.0",
    "handlebars": "^4.7.6",
    "stdout-stderr": "^0.1.13",
    "tmp": "^0.2.0"
  },
  "engines": {
    "node": ">= 12.*"
  },
  "files": [
    "/lib",
    "/static"
  ],
  "homepage": "https://github.com/checkupjs/checkup",
  "license": "MIT",
  "main": "lib/index.js",
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/checkupjs/checkup",
  "scripts": {
    "build": "tsc --build",
    "test": "jest --passWithNoTests --no-cache"
  },
  "types": "lib/index.d.ts"
}
