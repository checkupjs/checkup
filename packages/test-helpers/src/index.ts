export { stdout } from './stdout';
export { createTmpDir } from './tmp-dir';
export { testRoot } from './test-root';
export { getTaskContext } from './get-task-context';
export { filterPieChartDataForTest } from './validate-chart-results';
export { default as CheckupProject } from './checkup-fixturify-project';
export { default as EmberProject } from './ember-cli-fixturify-project';
export { default as Plugin } from './plugin';
