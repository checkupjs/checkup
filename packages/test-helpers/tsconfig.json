{
  "extends": "../../tsconfig-base.json",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "types": ["chai", "jest", "node"],
    "esModuleInterop": true
  },
  "references": [
    {
      "path": "../core"
    }
  ],
  "include": ["src/**/*"]
}
