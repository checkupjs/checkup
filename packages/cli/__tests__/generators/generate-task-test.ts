/* eslint-disable jest/expect-expect */

import * as helpers from 'yeoman-test';

import TaskGenerator from '../../src/generators/task';
import { generatePlugin } from '../__utils__/generate-plugin';
import { testRoot } from '@checkup/test-helpers';

function assertTaskFiles(name: string, dir: string, extension: string = 'ts') {
  let root = testRoot(dir);

  expect(root.file(`src/tasks/${name}-task.${extension}`).contents).toMatchSnapshot();
  expect(root.file(`src/results/${name}-task-result.${extension}`).contents).toMatchSnapshot();
  expect(root.file(`__tests__/${name}-task-test.${extension}`).contents).toMatchSnapshot();
}

function assertPluginFiles(dir: string, extension: string = 'ts') {
  let root = testRoot(dir);

  expect(root.file(`src/hooks/register-tasks.${extension}`).contents).toMatchSnapshot();
}

describe('task generator', () => {
  it('generates correct files with TypeScript for defaults', async () => {
    let baseDir = await generatePlugin();
    let dir = await helpers
      .run(TaskGenerator, { namespace: 'checkup:task' })
      .cd(baseDir)
      .withOptions({
        name: 'my-foo',
        defaults: true,
      });

    assertTaskFiles('my-foo', dir);
    assertPluginFiles(dir);
  });

  it('generates multiple correct files with TypeScript for defaults', async () => {
    let baseDir = await generatePlugin();
    let dir = await helpers
      .run(TaskGenerator, { namespace: 'checkup:task' })
      .cd(baseDir)
      .withOptions({
        name: 'my-foo',
        defaults: true,
      });

    dir = await helpers.run(TaskGenerator, { namespace: 'checkup:task' }).cd(baseDir).withOptions({
      name: 'my-bar',
      defaults: true,
    });

    assertTaskFiles('my-foo', dir);
    assertTaskFiles('my-bar', dir);
    assertPluginFiles(dir);
  });

  it('generates correct files with JavaScript', async () => {
    let baseDir = await generatePlugin(
      { name: 'my-plugin', defaults: false },
      { typescript: false }
    );
    let dir = await helpers
      .run(TaskGenerator, { namespace: 'checkup:task' })
      .cd(baseDir)
      .withOptions({
        name: 'my-foo',
      })
      .withPrompts({
        typescript: false,
      });

    assertTaskFiles('my-foo', dir, 'js');
    assertPluginFiles(dir, 'js');
  });

  it('generates correct files with category', async () => {
    let baseDir = await generatePlugin();
    let dir = await helpers
      .run(TaskGenerator, { namespace: 'checkup:task' })
      .cd(baseDir)
      .withOptions({
        name: 'my-foo',
      })
      .withPrompts({
        category: 'Insights',
      });

    assertTaskFiles('my-foo', dir);
    assertPluginFiles(dir);
  });

  it('generates correct files with priority', async () => {
    let baseDir = await generatePlugin();
    let dir = await helpers
      .run(TaskGenerator, { namespace: 'checkup:task' })
      .cd(baseDir)
      .withOptions({
        name: 'my-foo',
      })
      .withPrompts({
        priority: 'High',
      });

    assertTaskFiles('my-foo', dir);
    assertPluginFiles(dir);
  });
});
