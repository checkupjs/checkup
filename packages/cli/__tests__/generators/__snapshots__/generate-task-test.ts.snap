// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task generator generates correct files with JavaScript 1`] = `
"import { BaseTask, Category, Priority } from '@checkup/core';
import MyFooTaskResult from '../results/my-foo-task-result';

export default class MyFooTask extends BaseTask {
  meta = {
    taskName: 'my-foo',
    friendlyTaskName: 'My Foo',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run() {
    let result = new MyFooTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates correct files with JavaScript 2`] = `
"import { BaseTaskResult, ui } from '@checkup/core';

export default class MyFooTaskResult extends BaseTaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates correct files with JavaScript 3`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyFooTask from '../src/tasks/my-foo-task';
import MyFooTaskResult from '../src/results/my-foo-task-result';

describe('my-foo-task', () => {
  let checkupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const taskResult = await new MyFooTask({ path: checkupProject.baseDir }).run();

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const taskResult = await new MyFooTask({ path: checkupProject.baseDir }).run();

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates correct files with JavaScript 4`] = `
"import { Hook } from '@oclif/config';

import MyFooTask from '../tasks/my-foo-task';

const hook = async function({ context, tasks }) {
  tasks.registerTask(new MyFooTask(context));
};

export default hook;
"
`;

exports[`task generator generates correct files with TypeScript for defaults 1`] = `
"import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import MyFooTaskResult from '../results/my-foo-task-result';

export default class MyFooTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'my-foo',
    friendlyTaskName: 'My Foo',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    let result: MyFooTaskResult = new MyFooTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates correct files with TypeScript for defaults 2`] = `
"import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class MyFooTaskResult extends BaseTaskResult implements TaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates correct files with TypeScript for defaults 3`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyFooTask from '../src/tasks/my-foo-task';
import MyFooTaskResult from '../src/results/my-foo-task-result';

describe('my-foo-task', () => {
  let checkupProject: CheckupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates correct files with TypeScript for defaults 4`] = `
"import { Hook } from '@oclif/config';

import MyFooTask from '../tasks/my-foo-task';

const hook: Hook<'register-tasks'> = async function({ context, tasks }: any) {
  tasks.registerTask(new MyFooTask(context));
};

export default hook;
"
`;

exports[`task generator generates correct files with category 1`] = `
"import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import MyFooTaskResult from '../results/my-foo-task-result';

export default class MyFooTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'my-foo',
    friendlyTaskName: 'My Foo',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    let result: MyFooTaskResult = new MyFooTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates correct files with category 2`] = `
"import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class MyFooTaskResult extends BaseTaskResult implements TaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates correct files with category 3`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyFooTask from '../src/tasks/my-foo-task';
import MyFooTaskResult from '../src/results/my-foo-task-result';

describe('my-foo-task', () => {
  let checkupProject: CheckupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates correct files with category 4`] = `
"import { Hook } from '@oclif/config';

import MyFooTask from '../tasks/my-foo-task';

const hook: Hook<'register-tasks'> = async function({ context, tasks }: any) {
  tasks.registerTask(new MyFooTask(context));
};

export default hook;
"
`;

exports[`task generator generates correct files with priority 1`] = `
"import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import MyFooTaskResult from '../results/my-foo-task-result';

export default class MyFooTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'my-foo',
    friendlyTaskName: 'My Foo',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.High,
    },
  };

  async run(): Promise<TaskResult> {
    let result: MyFooTaskResult = new MyFooTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates correct files with priority 2`] = `
"import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class MyFooTaskResult extends BaseTaskResult implements TaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates correct files with priority 3`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyFooTask from '../src/tasks/my-foo-task';
import MyFooTaskResult from '../src/results/my-foo-task-result';

describe('my-foo-task', () => {
  let checkupProject: CheckupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates correct files with priority 4`] = `
"import { Hook } from '@oclif/config';

import MyFooTask from '../tasks/my-foo-task';

const hook: Hook<'register-tasks'> = async function({ context, tasks }: any) {
  tasks.registerTask(new MyFooTask(context));
};

export default hook;
"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 1`] = `
"import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import MyFooTaskResult from '../results/my-foo-task-result';

export default class MyFooTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'my-foo',
    friendlyTaskName: 'My Foo',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    let result: MyFooTaskResult = new MyFooTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 2`] = `
"import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class MyFooTaskResult extends BaseTaskResult implements TaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 3`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyFooTask from '../src/tasks/my-foo-task';
import MyFooTaskResult from '../src/results/my-foo-task-result';

describe('my-foo-task', () => {
  let checkupProject: CheckupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const result = await new MyFooTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyFooTaskResult>result;

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 4`] = `
"import { BaseTask, Category, Priority, Task, TaskMetaData, TaskResult } from '@checkup/core';
import MyBarTaskResult from '../results/my-bar-task-result';

export default class MyBarTask extends BaseTask implements Task {
  meta: TaskMetaData = {
    taskName: 'my-bar',
    friendlyTaskName: 'My Bar',
    taskClassification: {
      category: Category.Insights,
      priority: Priority.Low,
    },
  };

  async run(): Promise<TaskResult> {
    let result: MyBarTaskResult = new MyBarTaskResult(this.meta);

    return result;
  }
}"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 5`] = `
"import { BaseTaskResult, TaskResult, ui } from '@checkup/core';

export default class MyBarTaskResult extends BaseTaskResult implements TaskResult {
  toConsole() {
    ui.styledHeader(this.meta.friendlyTaskName);
  }

  toJson() {
    return {
      meta: this.meta,
      result: {},
    };
  }

  toReportData() {}
}
"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 6`] = `
"import { CheckupProject, stdout } from '@checkup/test-helpers';
import MyBarTask from '../src/tasks/my-bar-task';
import MyBarTaskResult from '../src/results/my-bar-task-result';

describe('my-bar-task', () => {
  let checkupProject: CheckupProject;

  beforeEach(() => {
    checkupProject = new CheckupProject('checkup-app', '0.0.0', project => {
      project.addDependency('ember-cli', '^3.15.0');
    });

    checkupProject.writeSync();
    checkupProject.gitInit();
  });

  afterEach(() => {
    checkupProject.dispose();
  });

  it('can read task and output to console', async () => {
    const result = await new MyBarTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyBarTaskResult>result;

    taskResult.toConsole();

    expect(stdout()).toMatchSnapshot();
  });

  it('can read task as JSON', async () => {
    const result = await new MyBarTask({ path: checkupProject.baseDir }).run();
    const taskResult = <MyBarTaskResult>result;

    expect(taskResult.toJson()).toMatchSnapshot();
  });
});"
`;

exports[`task generator generates multiple correct files with TypeScript for defaults 7`] = `
"import { Hook } from '@oclif/config';

import MyBarTask from '../tasks/my-bar-task';

import MyFooTask from '../tasks/my-foo-task';

const hook: Hook<'register-tasks'> = async function({ context, tasks }: any) {
  tasks.registerTask(new MyFooTask(context));
  tasks.registerTask(new MyBarTask(context));
};

export default hook;
"
`;
